File: python/exercises.py
from dataclasses import dataclass
from collections.abc import Callable
from typing import Tuple
import math


def change(amount: int) -> dict[int, int]:
    if not isinstance(amount, int):
        raise TypeError('Amount must be an integer')
    if amount < 0:
        raise ValueError('Amount cannot be negative')
    counts, remaining = {}, amount
    for denomination in (25, 10, 5, 1):
        counts[denomination], remaining = divmod(remaining, denomination)
    return counts


# Write your first then lower case function here
''' iterates over a string-- and checks if the item satisfies the predicate function'''
def first_then_lower_case(lst, predicate):
    for item in lst:
        if predicate(item):
            return item.lower()
        ''' if not, return none'''
    return None

# Write your powers generator here
''' generator function  yields powers of a base up to a specified limit '''
def powers_generator(base: int, limit: int):
    power = 1  
    while power <= limit:  
        yield power  
        '''multiply the current power by the base to get the next power'''
        power *= base 



# Write your say function here
# Function that accumulates words and returns them as a single space-separated string
def say(word=None):
    '''initialize an empty list to store the words'''
    words = [] 
    '''handle adding new words or returning the result'''
    def inner(new_word=None):
        if new_word is not None: 
            ''' add the new word to the list '''
            words.append(new_word)  
            return inner  
        '''check for new word if not return the accumulated words as a space-separated string'''
        return ' '.join(words)  
    
    if word is not None:  
        '''add initial word to list'''
        words.append(word)  
        return inner  
    '''if no initial word is provided, return the result of the inner function immediately'''
    return inner()  





# Write your line count function here
def meaningful_line_count(filename: str) -> int:
    try:
        with open(filename, 'r') as file:
            '''initialize a counter for meaningful lines'''
            count = 0  
            for line in file:
                '''remove front and end whitespace'''
                line = line.strip()  
                if line and not line.startswith('#'): 
                    count += 1  
                    '''' Return the final count of meaningful lines'''
            return count  
    except FileNotFoundError as e:  
        raise FileNotFoundError(f'No such file: {filename}') from e



# Write your Quaternion class here
@dataclass(frozen=True)
class Quaternion:
    a: float
    b: float
    c: float
    d: float

    def __add__(self, other: 'Quaternion') -> 'Quaternion':
        """
        Adds two quaternions, 
        returns sum
        """
        return Quaternion(
            self.a + other.a,
            self.b + other.b,
            self.c + other.c,
            self.d + other.d
        )

    def __mul__(self, other: 'Quaternion') -> 'Quaternion':
        """
        Multiplies two quaternions, 
        returns product
        """
        return Quaternion(
            self.a * other.a - self.b * other.b - self.c * other.c - self.d * other.d,
            self.a * other.b + self.b * other.a + self.c * other.d - self.d * other.c,
            self.a * other.c - self.b * other.d + self.c * other.a + self.d * other.b,
            self.a * other.d + self.b * other.c - self.c * other.b + self.d * other.a
        )

    def __eq__(self, other: object) -> bool:
        """
        Checks if two quaternions are equal
        returns true if both quaternions are equal, false otherwise
        """
        if not isinstance(other, Quaternion):
            return NotImplemented
        return (
            math.isclose(self.a, other.a) and
            math.isclose(self.b, other.b) and
            math.isclose(self.c, other.c) and
            math.isclose(self.d, other.d)
        )

    def __str__(self) -> str:
        """
        Returns a string representation of the quaternion in the form 'a + bi + cj + dk'
        """
        parts = []

        def format_term(value: float, letter: str) -> str:
            """
            Formats a term of the quaternion for string representation, and returns it
            """
            if math.isclose(value, 0):
                return ''
            if math.isclose(value, 1):
                return letter
            if math.isclose(value, -1):
                return '-' + letter
            return f"{value}{letter}"

        if not math.isclose(self.a, 0):
            parts.append(f"{self.a}")

        if not math.isclose(self.b, 0):
            parts.append(format_term(self.b, 'i'))

        if not math.isclose(self.c, 0):
            parts.append(format_term(self.c, 'j'))

        if not math.isclose(self.d, 0):
            parts.append(format_term(self.d, 'k'))

        if len(parts) == 0:
            return "0"

        result = parts[0]
        for part in parts[1:]:
            if part.startswith('-'):
                result += part
            else:
                result += '+' + part

        return result

    @property
    def coefficients(self) -> Tuple[float, float, float, float]:
        """
        Returns the coefficients of the quaternion as a tuple
        """
        return self.a, self.b, self.c, self.d

    @property
    def conjugate(self) -> 'Quaternion':
        """
        Returns the conjugate of the quaternion
        """
        return Quaternion(self.a, -self.b, -self.c, -self.d)

File: lua/exercises.lua
function change(amount)
  if math.type(amount) ~= "integer" then
    error("Amount must be an integer")
  end
  if amount < 0 then
    error("Amount cannot be negative")
  end
  local counts, remaining = {}, amount
  for _, denomination in ipairs({25, 10, 5, 1}) do
    counts[denomination] = remaining // denomination
    remaining = remaining % denomination
  end
  return counts
end

-- This function takes an array and a predicate function as arguments
-- It returns the first value that satisfies the predicate, converted to lowercase
-- If no value satisfies the predicate, it returns nil
function first_then_lower_case(array, predicate)
  -- Iterate through each value in the array
  for _, value in pairs(array) do
      -- Check if the current value satisfies the predicate function
      if predicate(value) then
          -- If it does, immediately return the lowercase version of this value
          return string.lower(value)
      end
  end
  -- If no value in the array satisfies the predicate, return nil
  return nil
end

-- This function creates a generator for powers of a given base up to a specified limit
-- It returns a coroutine that yields successive powers
function powers_generator(base, limit)
  -- Initialize the power to 1 (anything to the power of 0 is 1)
  local power = 1
  -- Create and return a new coroutine
  return coroutine.create(function()
      -- Continue generating powers while we're below or equal to the limit
      while power <= limit do
          -- Yield the current power value
          coroutine.yield(power)
          -- Calculate the next power by multiplying the current power by the base
          power = power * base
      end
  end)
end

-- This function creates a sentence builder
-- It takes a word as an argument and returns a function that can be used to build a sentence
function say(word)
  -- If no word is provided (nil), return an empty string
  if word == nil then
      return ""
  end
  
  -- Return a new function that takes the next word as an argument
  return function(next)
      -- If there's no next word, we've reached the end of the sentence
      if next == nil then
          -- Return the current word, effectively ending the sentence
          return word
      else
          -- If there is a next word, recursively call say() with the current word
          -- concatenated with a space and the next word
          return say(word .. " " .. next)
      end
  end
end

function meaningful_line_count(user_file)
  -- We open the file here
  local file = assert(io.open(user_file, "r"))
  -- Initialize a counter for meaningful lines
  local count = 0
  -- Read the file line by line
  for line in file:lines() do
      -- Trim leading and trailing whitespace
      line = line:match("^%s*(.-)%s*$")
      -- Check if the line is not empty and doesn't start with '#'
      if line ~= "" and line:sub(1, 1) ~= "#" then
          count = count + 1
      end
  end
  -- Close the file
  file:close()
  -- If count is 0, return an empty table, otherwise return the count
  return count > 0 and count or {}
end

-- quaternion class
Quaternion = {}
Quaternion.__index = Quaternion

-- constructor to create a new Quaternion
function Quaternion.new(a, b, c, d)
  local self = setmetatable({}, Quaternion)
  self.a = a
  self.b = b
  self.c = c
  self.d = d
  return self
end

-- overload the '+' operator for quaternion addition
function Quaternion.__add(q1, q2)
  return Quaternion.new(q1.a + q2.a, q1.b + q2.b, q1.c + q2.c, q1.d + q2.d)
end

-- overload the '*' operator for quaternion multiplication
function Quaternion.__mul(q1, q2)
  return Quaternion.new(
    q1.a * q2.a - q1.b * q2.b - q1.c * q2.c - q1.d * q2.d,
    q1.a * q2.b + q1.b * q2.a + q1.c * q2.d - q1.d * q2.c,
    q1.a * q2.c - q1.b * q2.d + q1.c * q2.a + q1.d * q2.b,
    q1.a * q2.d + q1.b * q2.c - q1.c * q2.b + q1.d * q2.a
  )
end

-- overload the '==' operator for quaternion equality check
function Quaternion.__eq(q1, q2)
  local epsilon = 1e-10
  return math.abs(q1.a - q2.a) < epsilon and
         math.abs(q1.b - q2.b) < epsilon and
         math.abs(q1.c - q2.c) < epsilon and
         math.abs(q1.d - q2.d) < epsilon
end

-- convert quaternion to string representation
function Quaternion.__tostring(q)
  local parts = {}

   -- helper function to format a term for string representation
  local function format_term(value, letter)
    if math.abs(value) < 1e-10 then
      return ''
    elseif math.abs(value - 1) < 1e-10 then
      return letter
    elseif math.abs(value + 1) < 1e-10 then
      return '-' .. letter
    else
      return value .. letter
    end
  end

   -- add real part if not zero
  if math.abs(q.a) >= 1e-10 then
    table.insert(parts, q.a)
  end

    -- add 'i', 'j', 'k' parts if not zero
  if math.abs(q.b) >= 1e-10 then
    table.insert(parts, format_term(q.b, 'i'))
  end

  if math.abs(q.c) >= 1e-10 then
    table.insert(parts, format_term(q.c, 'j'))
  end

  if math.abs(q.d) >= 1e-10 then
    table.insert(parts, format_term(q.d, 'k'))
  end

  -- return '0' if no parts were added, otherwise join the parts into a string
  if #parts == 0 then
    return "0"
  end

  local result = parts[1]
  for i = 2, #parts do
    if parts[i]:sub(1, 1) == '-' then
      result = result .. parts[i]
    else
      result = result .. '+' .. parts[i]
    end
  end

  return result
end

-- get the coefficients of the quaternion
function Quaternion:coefficients()
  return {self.a, self.b, self.c, self.d}
end

-- get the conjugate of the quaternion
function Quaternion:conjugate()
  return Quaternion.new(self.a, -self.b, -self.c, -self.d)
end
